### PR TITLE
perf(engine-server): disable reactivity in SSR

### PR DESCRIPTION
**Pull request by nolanlawson**
## Details

Removes reactivity (`observable-membrane`, `ReactiveObserver`) from `@lwc/engine-server`. Reactivity is not needed server-side.

The bundle size for `engine-server.js` drops from 246 kB to 218 kB (**27.8kB / 11.3% reduction**).

Performance in the `engine-server` Tachometer benchmarks improves **13%-18%**:

Screen Shot 2022-07-28 at 11 09 32 AM
Screen Shot 2022-07-28 at 11 09 36 AM

## Does this pull request introduce a breaking change?

* ✅ No, it does not introduce a breaking change.

**Technically** this would be a breaking change if any consumer were directly using `@lwc/engine-core`. They would need to know to transform `process.env.IS_BROWSER`. But I'm certain we don't have any consumers who are doing that.

## Does this pull request introduce an observable change?

* ✅ No, it does not introduce an observable change.

### CODE REVIEW

File: packages/@lwc/engine-core/src/framework/base-lightning-element.ts
Comment (nolanlawson) [inline packages/@lwc/engine-core/src/framework/base-lightning-element.ts:224]: Not directly related to reactivity, but we don't need this Locker expando in SSR.

File: packages/@lwc/engine-core/src/framework/mutation-tracker.ts
Comment (nolanlawson) [inline packages/@lwc/engine-core/src/framework/mutation-tracker.ts:22]: I tried to do this more elegantly in TypeScript, but it ended up being a huge hassle because TypeScript complains about missing private properties on the `ReactiveObserver`. Doing this `unknown` cast results in a lot less code.

File: packages/@lwc/engine-core/src/framework/decorators/api.ts
Comment (ravijayaramappa) [inline packages/@lwc/engine-core/src/framework/decorators/api.ts:156]: While going through this PR, I thought why does the `if (process.env.IS_BROWSER)` branching have to be in the utils(like mutation-tracker.ts) instead of letting the caller deciding if something needs to be reactive or not. But checking for references of the function shows there are many places that these functions are used and it can get complex if we start adding branching logic there. So for readability, it is best we branch it in the util.

Now for consistency, `AccessorReactiveObserver` should follow the same pattern as `createReactiveObserver` and let the dummy be stubbed out based on `process.env.IS_BROWSER`.

Comment (nolanlawson) [reply packages/@lwc/engine-core/src/framework/decorators/api.ts:156]: That sounds reasonable. Although I wonder if we could remove this code entirely instead? Do we ever plan to enable `ENABLE_REACTIVE_SETTER`?

Comment (nolanlawson) [reply packages/@lwc/engine-core/src/framework/decorators/api.ts:156]: OK, added `createAccessorReactiveObserver`.

**Review by ravijayaramappa — APPROVED** [review]
LGTM, prefer to get another +1 from @caridy or @pmdartus before merging.

File: packages/@lwc/engine-core/src/framework/mutation-tracker.ts
Comment (pmdartus) [reply packages/@lwc/engine-core/src/framework/mutation-tracker.ts:22]: The main concern that I have here is that the `DUMMY_REACTIVE_OBSERVER` is missing the `notify` method and TypeScript doesn't complain.

I think we can do better here without compromising on code size (yes, yes it's possible 😉). TypeScript complains because you are trying to tell TypeScript that `DUMMY_REACTIVE_OBSERVER` is an instance of `ReactiveObserver`. This is why it complains that private properties are missing.

The key here is to create an interface that would be implemented by both the `DUMMY_REACTIVE_OBSERVER` and the current `ReactiveObserver` class.  

```ts
export interface ReactiveObserver {
    observe(job: JobFunction): void;
    notify(): void;
    reset(): void;
    link(reactiveObservers: ReactiveObserver[]): void;
}

const DUMMY_REACTIVE_OBSERVER: ReactiveObserver = {  /* ... */ }

class ReactiveObserverImpl implements ReactiveObserver {}
```

Granted the `Impl` suffix really looks like Java, but it's the most straightforward way which guarantees type safety. The same observation applies to `AccessorReactiveObserver`.

File: packages/@lwc/engine-core/src/framework/decorators/api.ts
Comment (caridy) [inline packages/@lwc/engine-core/src/framework/decorators/api.ts:79]: I'm not a fan of having this here... this API is more suitable for the actual place where we implement it. Can we move this into that other file?

File: packages/@lwc/engine-core/src/framework/base-lightning-element.ts
Comment (caridy) [reply packages/@lwc/engine-core/src/framework/base-lightning-element.ts:224]: well, I think we do :) once we have locker for SSR jejejeje

File: packages/@lwc/engine-core/src/framework/membrane.ts
Comment (caridy) [inline packages/@lwc/engine-core/src/framework/membrane.ts:30]: this is the only piece that concerns me. The fact that in SSR the value passed down to you can be mutated it is a problem.

Comment (nolanlawson) [reply packages/@lwc/engine-core/src/framework/membrane.ts:30]: Hm, should we return a frozen copy instead?

File: packages/@lwc/engine-core/src/framework/mutation-tracker.ts
Comment (nolanlawson) [reply packages/@lwc/engine-core/src/framework/mutation-tracker.ts:22]: Great minds think alike! 🙂 This is exactly the code I started writing. But it was so much extra boilerplate that I wondered if it really added much value.

It especially got awkward because `ReactiveObserver` is imported from the file as both a class and a type, so consumers would have to know whether they were importing `ReactiveObserver` or `ReactiveObserverImpl`... it got messy fast.

File: packages/@lwc/engine-core/src/framework/membrane.ts
Comment (nolanlawson) [reply packages/@lwc/engine-core/src/framework/membrane.ts:30]: Hm, well I tried to write my own lightweight deep-frozen proxy, but it's not as fast as `observable-membrane`. 😅 So I guess for `getReadOnlyProxy`, we still need to defer to `observable-membrane` on both the client and the server.

The performance is still improved by 9%-12%, which is not as good as 13%-18%, but it's still better than 0.

Screen Shot 2022-07-29 at 3 17 08 PM
Screen Shot 2022-07-29 at 3 17 13 PM

I also added a test to confirm that children cannot mutate props passed in by their parents.

File: packages/@lwc/engine-core/src/framework/decorators/api.ts
Comment (nolanlawson) [reply packages/@lwc/engine-core/src/framework/decorators/api.ts]: Sure, moved `AccessorReactiveObserver` into its own file. (I didn't want to bloat `membrane.ts`.

File: packages/@lwc/engine-core/src/framework/membrane.ts
Comment (nolanlawson) [reply packages/@lwc/engine-core/src/framework/membrane.ts]: @caridy What do you think if maybe the server components could run in a kind of "validation mode" that would include the read-only proxy checks, but in "prod mode", the read-only proxy would be disabled?

There are competing desires here: we _should_ disallow mutations on the properties, especially because it matches the client-side behavior, but if you want your SSR to perform as well as possible, and if all your components are well-behaved, then it's just extra overhead. ~6% can be a big deal in terms of server throughput.

I suppose we could make the same argument for the client-side, e.g. only doing the read-only proxy for `process.env.NODE_ENV === 'development'`.

Comment (caridy) [reply packages/@lwc/engine-core/src/framework/membrane.ts]: Na', I think we should stay away from that kind of non-deterministic behavior mechanism based on the environment.

**Review by caridy — APPROVED** [review]

**Pull request merged**
